libdnf/advisory/advisory.cpp

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/advisory/advisory.hpp"
+
+#include "libdnf/advisory/advisory_collection.hpp"
+#include "libdnf/advisory/advisory_reference.hpp"
+#include "libdnf/common/exception.hpp"
+#include "libdnf/rpm/solv_sack_impl.hpp"
+#include "libdnf/utils/utils_internal.hpp"
+
+#include <fmt/format.h>
+#include <solv/pool.h>
+
+namespace libdnf::advisory {
+
+Advisory::Advisory(libdnf::rpm::SolvSack & sack, AdvisoryId id) : id(id), sack(sack.get_weak_ptr()) {}
+
+std::string Advisory::get_name() const {
+    const char * name;
+    Pool * pool = sack->p_impl->get_pool();
+
+    name = pool_lookup_str(pool, id.id, SOLVABLE_NAME);
+
+    if (strncmp(
+            libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX, name, libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH) !=
+        0) {
+        auto msg = fmt::format(
+            R"**(Bad libsolv id for advisory "{}", solvable name "{}" doesn't have advisory prefix "{}")**",
+            id.id,
+            name,
+            libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX);
+        throw RuntimeError(msg);
+    }
+
+    return std::string(name + libdnf::utils::SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH);
+}
+
+Advisory::Type Advisory::get_type() const {
+    const char * type;
+    Pool * pool = sack->p_impl->get_pool();
+    type = pool_lookup_str(pool, id.id, SOLVABLE_PATCHCATEGORY);
+
+    if (type == NULL) {
+        return Advisory::Type::UNKNOWN;
+    }
+    if (!strcmp(type, "bugfix")) {
+        return Advisory::Type::BUGFIX;
+    }
+    if (!strcmp(type, "enhancement")) {
+        return Advisory::Type::ENHANCEMENT;
+    }
+    if (!strcmp(type, "security")) {
+        return Advisory::Type::SECURITY;
+    }
+    if (!strcmp(type, "newpackage")) {
+        return Advisory::Type::NEWPACKAGE;
+    }
+
+    return Advisory::Type::UNKNOWN;
+}
+
+const char * Advisory::advisory_type_to_cstring(Type type) {
```

Comment:
(For a discussion) The type is defined as bit in int. In case that multiple bits are set we will be unable to return correct values.

Reply:
This is also related to https://github.com/rpm-software-management/libdnf/pull/1043#discussion_r625916499.